RecoLocalTracker/SiPixelClusterizer/plugins/gpuClusterChargeCut.h

```diff
@@ -0,0 +1,125 @@
+#ifndef RecoLocalTracker_SiPixelClusterizer_plugins_gpuClusterChargeCut_h
+#define RecoLocalTracker_SiPixelClusterizer_plugins_gpuClusterChargeCut_h
+
+#include <cstdint>
+#include <cstdio>
+
+#include "HeterogeneousCore/CUDAUtilities/interface/cuda_assert.h"
+#include "HeterogeneousCore/CUDAUtilities/interface/prefixScan.h"
+
+#include "gpuClusteringConstants.h"
+
+namespace gpuClustering {
+
+  __global__ void clusterChargeCut(
+      uint16_t* __restrict__ id,                 // module id of each pixel (modified if bad cluster)
+      uint16_t const* __restrict__ adc,          //  charge of each pixel
+      uint32_t const* __restrict__ moduleStart,  // index of the first pixel of each module
+      uint32_t* __restrict__ nClustersInModule,  // modified: number of clusters found in each module
+      uint32_t const* __restrict__ moduleId,     // module id of each module
+      int32_t* __restrict__ clusterId,           // modified: cluster id of each pixel
+      uint32_t numElements) {
+    if (blockIdx.x >= moduleStart[0])
+      return;
+
+    auto firstPixel = moduleStart[1 + blockIdx.x];
+    auto thisModuleId = id[firstPixel];
+    assert(thisModuleId < MaxNumModules);
+    assert(thisModuleId == moduleId[blockIdx.x]);
+
+    auto nclus = nClustersInModule[thisModuleId];
+    if (nclus == 0)
+      return;
+
+    if (threadIdx.x == 0 && nclus > MaxNumClustersPerModules)
+      printf("Warning too many clusters in module %d in block %d: %d > %d\n",
+             thisModuleId,
+             blockIdx.x,
+             nclus,
+             MaxNumClustersPerModules);
```

Comment:
@slava77 I don't know of any ways of passing asynchronous messages from a CUDA kernel back to the host (e.g. to the MessageLogger) - `printf` is a black box reimplemented by NVIDIA, without any way of redirecting its output.

SYCL implements an `ostream`-like class (see e.g. https://developer.codeplay.com/products/computecpp/ce/guides/sycl-guide/debugging), although it's hardcoded to eventually print to `cout`, so I guess we could eventually try to extract that functionality and reimplement something similar for the `MessageLogger` ?

Anyway, for the time being the only option is `printf`.

Reply:
just to clarify my proposal
- add `if (debug` was meant to allow suppression of the message if it is more useful for debuging or at the time/point of execution
- `pass the check/condition outside of the kernel to be picked up by the MessageLogger` meant to propose to add an output of the algorithm/kernel to fill warnings/summary counters or more complex details which can then be parsed by the CPU caller code to provide aggregated LogWarning/LogError message. This should also allow to avoid possible flooding of the logs.

So, IMHO, printf is not the only option.

Reply:
Yes, `if (debug)` is clearly a possibility - that still uses `printf`,  and the main question is what messages should be suppressable and which ones should not be.

While this proposal
> add an output of the algorithm/kernel to fill warnings/summary counters or more complex details which can then be parsed by the CPU caller code to provide aggregated LogWarning/LogError message

IMHO has at least two problems:
  - it requires a completely new mechanism to pass information back from the devices to the host, which would add some overhead also to the kernels that don't generate any messages (at the very least the extra buffers and copies to the host);
  - it implies a separation of where the conditions are checked (in the kernel) and where the messages are output (in the EDM modules), which doesn't look like a good solution from the encapsulation point of view.

I'm not saying it's not doable or that we should not pursue this - but it will require some prototyping, benchmarking, review, *etc.*, which IMHO is outside the scope of this PR.
If anything, it could happen in CMSSW after the current PRs have been merged, or in the Patatrack branch in parallel to the ongoing review, if anybody is interested in exploring it.

Reply:
An "output" of the algorithm/kernel to fill warnings/summary counters or more complex details which, by some expensive mechanism, can be passed back event by event is implemented in the "CA".
The current implementation does not work in case of multiple GPU...

Reply:
Many of these mechanisms are currently off by default as they negatively affect performance

Reply:
I would also like to stress that the goal is to stay on GPU as much as possible, move to HOST the very minimal possibly in bulk transfers.

Reply:
printf is not supposed to happen: 1) slows down enormously the process, 2) is limited to 1MB buffer per kernel (never managed to configure it larger)

Reply:
strictly speaking this very condition is one of the few that can indeed be checked and LogWarned in the module as the number of clusters is moved to host prior to HitBulding...

Reply:
> is limited to 1MB buffer per kernel (never managed to configure it larger)

You should be able to set a 16MB buffer with
```
process.CUDAService.limits.cudaLimitPrintfFifoSize = 16*1024*1024
```

The minimum if 512k, I did not investigate what is the maximum.

Reply:
Once upon a time, with Matti's help, I did that. No effect. Maybe with more recent CUDA we can give another try...

Reply:
From the comments, am I correct to conclude that a very abbreviated passing/return of nWarnings or nErrors per kernel call (atomic-added across the threads) would still be cheap enough to both collect on GPU and to transfer to HOST?

Reply:
For the time being "cheap enough" has not been established

Reply:
transfer to HOST even module by module is not cheap

Reply:
accumulate on GPU and transfer to HOST at end of job is at the moment not-multi-GPU safe

Reply:
transferring at the end of the job sounds like a reasonable case only when some 99%+ of everything stays on a GPU (meaning all event data processing stays on GPU).
I'm not sure we are anywhere close to that point.

Anyways, doesn't framework have to know for every module per event if a call to aquire completed? Is there a way to start from this knowledge and have a bit passed there that there was some warning? If yes, it could then be used as a trigger for the ::produce side to extract the warning/error data from the device and issue a relevant Log message.

Reply:
> Anyways, doesn't framework have to know for every module per event if a call to aquire completed?

This is done with a callback function queued in the CUDA stream with `cudaStreamAddCallback()` (to be changed possibly to a separate thread waiting on `cudaStreamSynchronize()` or `cudaEventSynchronize()` when `cudaStreamAddCallback()` gets removed from CUDA). 

> Is there a way to start from this knowledge and have a bit passed there that there was some warning?

That (or any of the `* Synchronize()`) provide no mean to pass any information from device to host (beyond an error code of CUDA itself). Passing arbitrary information implies an explicit transfer, implicit transfer with managed memory, or direct write host memory (via unified memory). 

Written that, the clusterizer already tranfers 2x 4 bytes (number of modules and number of clusters) from device to host, and those are dominated by overheads, and therefore increasing the size of the transfers is basically free. With some effort it could be possible to try to make only one transfer, perhaps some per-event counting information on warnings could be included there (memory allocations and counting itself will add some cost even if the transfer becomes free). But event then I'd argue that would be best done in its own PR.

Reply:
removal of `cout`s in production setup is a generally rather strong requirement in all our (CMSSW, not just reco) code reviews. The leading argument, IIRC, is the thread safety; other than lack of ability to trace the origin of the message.

Is `printf` here somehow thread safe?
If it has the same/similar issues as `cout`, then I'd expect at least silencing for a default (matrix workflow) setup still to be done in this PR.

Reply:
The `printf` most certainly has the same issues as `cout` (the use of stdout itself is thread safe, but the output may be garbled).

Reply:
`printf` in device code is definitely thread safe.
As far as I can tell, the output from multiple GPU threads is handled properly for each individual `printf` call; so if for example each call to `printf` terminated with a `\n` they will end up each on a separate line (unlike `cout`).

Silencing **error conditions** by default seems like the best way to never find out about them.
It probably makes sense for `LogInfo`, but if `LogWarning` and `LogError` are silenced by default, why do we have them at all ?

Reply:
IMHO a proper error condition is the one that is sent/managed via a central messaging service.

I'm learning through this thread that the formatting is apparently reliable; this is a positive in support of still keeping these cuda printfs.
Do they have a way to mix appropriately with MessageLogger messages or should we expect the output characters intermixed?

I'm open to discussion for keeping these printfs still enabled in the default workflows, but I still think that in each module/kernel there has to be a way to turn them off (just like in the MessageLogger I can silence messages by level/category/module).

Reply:
> Do they [cuda `printf`] have a way to mix appropriately with MessageLogger messages or should we expect the output characters intermixed?

My guess is that
  - all CUDA `printf` messages from one kernel will appear only after the kernel that originated them returns
  - the CUDA `printf` can be interspersed with the MessageLogger messages just like a single host `printf` could
  - I do not know whether that applies to the individual `printf` or to a kernel as a whole

If going through the MessageLogger is a requirement, somebody will have to develop a device-side interface.

Barring that, the only kind of integration I can think of is
  - each kernel is given a `verbose` parameter (either a boolean or an integer with a level, if we want to be overly fancy, or a bitmask with different flags for each level...)
  - each EDProducer queries the MessageLogger to check if its messages should be enabled (globally, or for a certain level / for each level), and passes this information to the kernels via the `verbose` flag

Reply:
can we move this discussion to an issue by itself?
After all it affects all GPU code, not just pixel local reco.

Anyhow I use very successfully printf in device code with multiple host threads to generate a csv file.
it is clobbered by cout and MessageLogger (or at least by the begin of event line) but absolutely thread safe by itself

Reply:
Moved to https://github.com/cms-sw/cmssw/issues/32502 .